cdc: add roachtest for azure event hub kafka support

wenyihu6 opened the pull request:
**cdc: add roachtest for azure event hub kafka support**

This patch adds a roachtest to verify that we are able to stream to Azure Event
Hubs via their Kafka endpoint.

To do this in the roachtests, the following was created in our CRL Azure org
(owned by test-eng):
- Resource group: e2e-infra-event-hub-rg
- Event Hubs Namespace: cdc-roachtest
- Event Hub: testing (within the "cdc-roachtest" namespace), serving as the
equivalent to Kafka topics.
- SAS policy: cdc-roachtest-auth-rule

The necessary credential env vars have been added to TeamCity agents by dev-inf.
Note that running this test on roachprod would not work due to lacking the
required credentials env vars set up.

The test involves the following steps:
1. Install Azure CLI on the Kafka node.
2. Retrieve the connection string for the Azure Event Hub "cdc-roachtest."
3. Create a changefeed to stream to the Azure Event Hub, followed by running a
brief workload for 30 minutes.

Resolves: https://github.com/cockroachdb/cockroach/issues/100815

----

**roachtest/cdc: rename kafkaManager.nodes to kafkaManager.kafkaSinkNode**

This patch renames the `nodes` field of kafkaManager to `kafkaSinkNode` to
avoid confusion as there has been only one kafka node in roachtest/cdc set up.

Epic: None
Release Note: None

samiskin committed:
changefeedccl: add azure event hub roachtest

This patch adds a roachtest to verify that we are able to stream to Azure Event
Hubs via their Kafka endpoint.

To do this in the roachtests, the following was created in our CRL Azure org
(owned by test-eng):
- Resource group: e2e-infra-event-hub-rg
- Event Hubs Namespace: cdc-roachtest
- Event Hub: testing (within the "cdc-roachtest" namespace), serving as the
equivalent to Kafka topics.
- SAS policy: cdc-roachtest-auth-rule

The necessary credential env vars have been added to TeamCity agents by dev-inf.
Note that running this test on roachprod would not work due to lacking the
required credentials env vars set up.

The test involves the following steps:
1. Install Azure CLI on the Kafka node.
2. Retrieve the connection string for the Azure Event Hub "cdc-roachtest."
3. Create a changefeed to stream to the Azure Event Hub, followed by running a
brief workload for 30 minutes.

Resolves: #100815

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/116950)
<!-- Reviewable:end -->

wenyihu6 commented:
Confirmed that the test passes on TeamCity agent. 

https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyAzureBazel/13386088?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true

<img width="875" alt="Screenshot 2024-01-03 at 3 32 34 PM" src="https://github.com/cockroachdb/cockroach/assets/56973754/9a6dc301-5b4a-4721-8d2f-44c0acd5c987">

srosenberg commented:
> Confirmed that the test passes on TeamCity agent.
https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyAzureBazel/13386088?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true

Looks like there was an authentication error?

> 22:18:22 test_runner.go:466: [w0] Creating new cluster for test cdc/kafka-azure: n2cpu4 (arch="amd64")
22:19:00 test_runner.go:678: [w0] Unable to create (or reuse) cluster for test cdc/kafka-azure due to: in provider: azure: resource group for location "eastus": resources.GroupsClient#Get: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="AuthorizationFailed" Message="The client 'c2b40df7-4cd7-4aed-bb74-00daabebec11' with object id 'c2b40df7-4cd7-4aed-bb74-00daabebec11' does not have authorization to perform action 'Microsoft.Resources/subscriptions/resourcegroups/read' over scope '/subscriptions/73075e4f-150d-4f4f-956b-053507b61fa7/resourcegroups/roachprod-vnets-eastus' or the scope is invalid. If access was recently granted, please refresh your credentials.".
22:19:00 github.go:290: skipping GitHub issue posting (not a release branch: "116950")

wenyihu6 commented:
> > Confirmed that the test passes on TeamCity agent.
> > https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyAzureBazel/13386088?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true
> 
> Looks like there was an authentication error?
> 
> > 22:18:22 test_runner.go:466: [w0] Creating new cluster for test cdc/kafka-azure: n2cpu4 (arch="amd64")
> > 22:19:00 test_runner.go:678: [w0] Unable to create (or reuse) cluster for test cdc/kafka-azure due to: in provider: azure: resource group for location "eastus": resources.GroupsClient#Get: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="AuthorizationFailed" Message="The client 'c2b40df7-4cd7-4aed-bb74-00daabebec11' with object id 'c2b40df7-4cd7-4aed-bb74-00daabebec11' does not have authorization to perform action 'Microsoft.Resources/subscriptions/resourcegroups/read' over scope '/subscriptions/73075e4f-150d-4f4f-956b-053507b61fa7/resourcegroups/roachprod-vnets-eastus' or the scope is invalid. If access was recently granted, please refresh your credentials.".
> > 22:19:00 github.go:290: skipping GitHub issue posting (not a release branch: "116950")

Do you mind posting a link for this log? I'm having trouble finding it.

srosenberg commented:
> Do you mind posting a link for this log? I'm having trouble finding it. I will try rerunning on gce worker and see if it works.

[test_runner.log](https://teamcity.cockroachdb.com/repository/download/Cockroach_Nightlies_RoachtestNightlyAzureBazel/13386088:id/test_runner-1704320302.log)

wenyihu6 commented:
> > Do you mind posting a link for this log? I'm having trouble finding it. I will try rerunning on gce worker and see if it works.
> 
> [test_runner.log](https://teamcity.cockroachdb.com/repository/download/Cockroach_Nightlies_RoachtestNightlyAzureBazel/13386088:id/test_runner-1704320302.log)

Confirmed that it worked on gce worker and running `./dev build roachtest && roachtest run cdc/kafka-azure --cockroach=./artifacts/cockroach`. I think this is likely because the client_id and client_secret I asked #dev-inf to set up doesn't have the proper permission to create cluster.  I will reach out to see if we should add the permission or seek alternatives.

wenyihu6 commented:
> > > Do you mind posting a link for this log? I'm having trouble finding it. I will try rerunning on gce worker and see if it works.
> > 
> > 
> > [test_runner.log](https://teamcity.cockroachdb.com/repository/download/Cockroach_Nightlies_RoachtestNightlyAzureBazel/13386088:id/test_runner-1704320302.log)
> 
> Confirmed that it worked on gce worker an `./dev build roachtest && roachtest run cdc/kafka-azure --cockroach=./artifacts/cockroach`. I think this is likely because the client_id and client_secret I asked #dev-inf to set up doesn't have the proper permission to create cluster over our resource. I will reach out to see if we should add the permission or seek alternatives.

Changed the env vars and should be fixed now.

wenyihu6 commented:
Friendly ping on the review here : )

srosenberg commented on pkg/cmd/roachtest/tests/cdc.go:
```diff
@@ -2120,6 +2187,62 @@ func (k kafkaManager) installJRE(ctx context.Context) error {
+	// az login --service-principal -t <Tenant-ID> -u <Client-ID> -p=<Client-secret>
+	cmdStr := fmt.Sprintf("az login --service-principal -t %s -u %s -p=%s", azureTenantID, azureClientID, azureClientSecret)
+	results, err := k.c.RunWithDetailsSingleNode(ctx, k.t.L(), k.nodes, cmdStr)
+	k.t.L().Printf("az login%s\n", results.Stdout)
```
I think we should consider using `redact.RedactableString` in order to redact secrets from stdout [1]. While the test environment doesn't have _real_ secrets (i.e., devoid of customer data), each run and its artifacts are _publicly_ accessible. So, in theory a bad actor could infiltrate the test account if all auth. secrets are logged in plaintext.

[1] https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/1824817806/Log+and+error+redactability

wenyihu6 replied:
I don't think the output from az login doesn't contain any secrets. But the output from `az account set` and `az eventhubs namespace` does, so I am intentionally now printing their outputs. I will slack you to confirm if the output from az login contains any secrets (not doing it here to avoid secret leak on github).

wenyihu6 replied:
I think I will just remove this output from our log to make us feel safer. The error message here seems to be sufficient. Just pushed the changes.

srosenberg commented on pkg/cmd/roachtest/tests/cdc.go:
```diff
@@ -245,6 +252,27 @@ func (ct *cdcTester) setupSink(args feedArgs) string {
+		kafka := kafkaManager{
+			t:     ct.t,
+			c:     ct.cluster,
+			nodes: kafkaNode,
```
Nit: might be more descriptive to call it `kafkaSinkNode`. Otherwise, at the callsite (of `RunWithDetailsSingleNode`) it's unclear whether it's a single-element slice.

wenyihu6 replied:
I agree with the confusion here. I was also bothered by it. I'm naming it as kafkaNode since this has been the [convention](https://github.com/cockroachdb/cockroach/blob/404376662fea066cb6f1968d5a1ecfb3c09ae467/pkg/cmd/roachtest/tests/cdc.go#L237) in cdc roachtests. But we should probably use the field name node rather than nodes since it is always a single node for kafka in roachtests/cdc. I will open another PR if you think this is a good idea.

srosenberg replied:
It's really a "nit", so it doesn't really merit another PR, imo. You could just `Refactor->Rename` here.

wenyihu6 replied:
I'm not sure if I'm interpreting the comment here right - why would renaming kafkaNode to kafkaSinkNode improve clarity regarding single-element slice. I thought it would make more sense to rename the field name `nodes` of kafkaManager to `node` to emphasize the singularity. Using kafkaNode seems to be the naming [convention](https://github.com/cockroachdb/cockroach/blob/404376662fea066cb6f1968d5a1ecfb3c09ae467/pkg/cmd/roachtest/tests/cdc.go) for all roachtests, so I thought just renaming it in this PR might lead to inconsistencies.

srosenberg replied:
> why would renaming kafkaNode to kafkaSinkNode

Indeed, that's not what the comment suggested, rather the `nodes` field which is plural.

wenyihu6 replied:
Ahh I see. Sorry for the misunderstanding here. I added the changes as a second commit as part of this PR.

srosenberg commented on pkg/cmd/roachtest/tests/cdc.go:
```diff
@@ -1598,6 +1626,33 @@ func registerCDC(r registry.Registry) {
+		Name:             "cdc/kafka-azure",
+		Owner:            `cdc`,
+		CompatibleClouds: registry.AllExceptAWS,
+		Cluster:          r.MakeClusterSpec(2, spec.Arch(vm.ArchAMD64), spec.GCEZones("us-east1-b")),
```
Does this test have a dependency to run on amd64 VMs only? If not, then it's best to leave it unspecified; this way it would be eligible for a metamorphic run on arm64.

wenyihu6 replied:
Agreed. Done.

srosenberg commented on pkg/cmd/roachtest/tests/cdc.go:
```diff
@@ -2120,6 +2187,62 @@ func (k kafkaManager) installJRE(ctx context.Context) error {
 
+// installAzureCli installs azure cli on the kafka node which is necessary for
+// retrieving endpoint connection string for the roachtest azure event hub.
+func (k kafkaManager) installAzureCli(ctx context.Context) error {
```
I wonder if we could simplify things. We already have `azure-cli` [1] on the node where (cdc) `Run` function is executing. Wouldn't be simpler to get the connection string there, say dump them to a flat file, and `roachprod.Put` to the kafka node?

[1] https://github.com/cockroachdb/cockroach/pull/105788

wenyihu6 replied:
Is the only benefit here to avoid installing azure cli on the node? The connection string has the secret key; dumping it to a flat file and `roachprod.Put` feels a bit more complicated.

srosenberg replied:
It's just adding a redundancy wrt boilerplate; maybe we could refactor it as a `roachprod install` command [1]. Otherwise, it's fine to keep the existing logic.

[1] https://github.com/cockroachdb/cockroach/blob/f428c197547b2dbf4191a0ffbe6f8f885d5329d9/pkg/roachprod/install/install.go#L23

srosenberg commented:
> Friendly ping on the review here : )

Only a few comments on simplifying bootstrapping and secrete management, otherwise looks good!

wenyihu6 committed:
roachtest/cdc: rename kafkaManager.nodes to kafkaManager.kafkaSinkNode

This patch renames the `nodes` field of kafkaManager to `kafkaSinkNode` to
avoid confusion as there has been only one kafka node in roachtest/cdc set up.

Epic: None
Release Note: None

# Conflicts:
#	pkg/cmd/roachtest/tests/cdc.go
